### PR TITLE
Change speed accuracy scaling to be closer to worst case scenario

### PR DESCRIPTION
**Pull request by emu1337**
The purpose of this change is to more accurately evaluate scores on maps where a lot of filler notes are present, which are trivial to accurately hit, but increase the overall accuracy.
Since the current implementation uses this overall accuracy to scale speed pp, it can create situations where the part of the map that grants all of the speed SR can be played poorly, but it would still reward the player because their overall accuracy isn't that bad in the eyes of the system.

To remedy this, a new attribute has been added, that is meant to keep information about how many notes in the map are actually difficult to hit accurately.
A new accuracy value is calculated for speed which assumes that all of the 100s/50s/misses are on these difficult notes. This is then used to scale the final pp value.

Example of a targeted map:
[Yuyoyuppe - Hidamari no Uta [Expert] (Seto Kousuke)](https://osu.ppy.sh/beatmapsets/831242#osu/1741498)
HDDT SS: 877pp -> 877pp
HDDT 94%: 656pp -> **423pp**

Example unaffected map due to lack of any filler notes:
[Various Artist - Long Stream Practice Maps [Neurotoxin BPM180] (onlyforyou)](https://osu.ppy.sh/beatmapsets/72474#osu/207692)
HDDT SS: 777pp -> 777pp
HDDT 95%: 605pp -> 605pp

### **TL;DR: hidamari nerf**

---

Spreadsheet of SR/PP changes: https://docs.google.com/spreadsheets/d/11CPI18QjRMpm0MkLGtdxKgne3g-LxWHIK2isxItQDRo/edit?usp=sharing

As of a7de43ade6459fe63d85f75d40cb7ea08b7d7e81

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs [hunk 1]
```diff
@@ -170,7 +174,17 @@ protected override double StrainValueAt(DifficultyHitObject current)
...
+
+            objectStrains.Add(totalStrain);
+
+            return totalStrain;
+        }
+
+        public double RelevantNoteCount()
+        {
+            double maxStrain = objectStrains.Max();
+            return objectStrains.Aggregate((total, next) => total + (1.0 / (1.0 + Math.Exp(-(next/maxStrain * 12.0 - 6.0)))));
```
[the remark this answers is not in the document]
Comment (stanriders) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs:187]: I'm referring to the fact that you count stream notes by taking combined finalized strains instead of doing that in the middle of processing, for example when the speed bonus happens

[the remark this answers is not in the document]
Comment (emu1337) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs:187]: It's done like this because the aggregation depends on `maxStrain` which is unknown during processing. I don't know if that term could be moved out of the exponential or not, thats beyond my math knowhow.

[the remark this answers is not in the document]
Comment (stanriders) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs:187]: What I meant is pretty much to use objects below `single_spacing_threshold` (and maybe with `speed bonus`?) instead of taking finals strains and guessing which ones are to be considered streams. That would also fix rhythmically-hard objects from being considered streams and potentially getting nerfed

[the remark this answers is not in the document]
Comment (emu1337) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs:187]: The change is meant to affect rhythmically-hard objects as well, not just streams. The goal is to scale accuracy more for the hardest 'speed' patterns, since the hardest pattern is what contributes to nearly all of the speed rating. You can't really expect to receive the full rewards if you cheesed the only hard section.

With that in mind, "guessing" by strain is the most straightforward way to count the number of hard objects in the map, because strain IS difficulty. Using only distance or deltatime doesn't tell you about the previous notes.
As an example, imagine a map with a bunch of easy triples and a long stream at the end. Those triples would count as relevant notes because they have low deltatime, and their distances are under the spacing threshold. If you completely cheesed the long stream, it wouldn't matter because the triples inflated the relevant note count. Here is such a [map](https://osu.ppy.sh/beatmapsets/369354#osu/809469) from the ranked pool.

**Review by stanriders — DISMISSED** [review]
I'm fine with it the way it is now, we can get it better later after splitting speed into aim and tap parts. For now it might be overestimating the difficulty, but that's better than what we have on live.

**Comment by apollo-dw** [comment]
I think we can progress on this one - will need a merge though. Keen on getting rid of the 50s nerf bandaid. I'm also leaning towards splitting the `RelevantNoteCount` function from `Speed` into it's own counting skill, but that could be over the top... @stanriders what do you think?

**Comment by stanriders** [comment]
I dont really see any reason to, why would you want it to be a separate skill considering it only affects speed?

**Comment by smoogipoo** [comment]
Test failures need attention:
```
  ----> System.InvalidOperationException : Sequence contains no elements
  Stack Trace:
  --TearDown
   at osu.Framework.Testing.TestScene.checkForErrors()
   at osu.Framework.Testing.TestScene.RunTestsFromNUnit()
--InvalidOperationException
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.Max(IEnumerable`1 source)
   at osu.Game.Rulesets.Osu.Difficulty.Skills.Speed.RelevantNoteCount() in D:\a\osu\osu\osu.Game.Rulesets.Osu\Difficulty\Skills\Speed.cs:line 59
   at osu.Game.Rulesets.Osu.Difficulty.OsuDifficultyCalculator.CreateDifficultyAttributes(IBeatmap beatmap, Mod[] mods, Skill[] skills, Double clockRate) in D:\a\osu\osu\osu.Game.Rulesets.Osu\Difficulty\OsuDifficultyCalculator.cs:line 41
```
NaN value in pp:
```
   ----> System.ArgumentException : Size must be finite, but is (NaN, 1).
  Stack Trace:
  --TearDown
   at osu.Framework.Testing.TestScene.checkForErrors()
   at osu.Framework.Testing.TestScene.RunTestsFromNUnit()
--ArgumentException
   at osu.Framework.Graphics.Drawable.set_Size(Vector2 value)
   at osu.Framework.Graphics.Transforms.Transform`3.Apply(Double time)
   at osu.Framework.Graphics.Transforms.Transformable.AddTransform(Transform transform, Nullable`1 customTransformID)
   at osu.Framework.Graphics.TransformableExtensions.TransformTo[TThis](TThis t, Transform transform)
   at osu.Framework.Graphics.TransformableExtensions.ResizeTo[T,TEasing](T drawable, Vector2 newSize, Double duration, TEasing& easing)
   at osu.Framework.Graphics.TransformableExtensions.ResizeTo[T](T drawable, Vector2 newSize, Double duration, Easing easing)
   at osu.Game.Graphics.UserInterface.Bar.updateBarLength() in D:\a\osu\osu\osu.Game\Graphics\UserInterface\Bar.cs:line 86
   at osu.Game.Graphics.UserInterface.Bar.set_Length(Single value) in D:\a\osu\osu\osu.Game\Graphics\UserInterface\Bar.cs:line 35
   at osu.Game.Screens.Ranking.Statistics.PerformanceBreakdownChart.createAttributeRow(PerformanceDisplayAttribute attribute, PerformanceDisplayAttribute perfectAttribute) in D:\a\osu\osu\osu.Game\Screens\Ranking\Statistics\PerformanceBreakdownChart.cs:line 205
   at osu.Game.Screens.Ranking.Statistics.PerformanceBreakdownChart.setPerformanceValue(PerformanceBreakdown breakdown) in D:\a\osu\osu\osu.Game\Screens\Ranking\Statistics\PerformanceBreakdownChart.cs:line 176
   at osu.Game.Screens.Ranking.Statistics.PerformanceBreakdownChart.<>c__DisplayClass17_0.<load>b__1() in D:\a\osu\osu\osu.Game\Screens\Ranking\Statistics\PerformanceBreakdownChart.cs:line 153
```

**Comment by smoogipoo** [comment]
Added spreadsheet of SR/PP changes to description.

**Comment by smoogipoo** [comment]
Awaiting for 2 approvals from @ppy/osu-pp-committee before merge. I can fix test failures if you’re unable to do so.

**Review by apollo-dw — DISMISSED** [review]

**Review by smoogipoo — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Comment by MashiroCodfish** [comment]
ppy我草泥马 你毁了多少人的成绩
